components/list/list.js

```diff
@@ -64,13 +71,65 @@ class List extends LitElement {
 		return [layout, specialCases];
 	}
 
+	firstUpdated(changedProperties) {
+		super.firstUpdated(changedProperties);
+		this._checked = {};
+		this.addEventListener('d2l-list-item-checked', this._onlistItemChecked.bind(this));
+	}
+
+	itemsChecked() {
+		return Object.keys(this._checked).filter(ref => this._checked[ref]);
+	}
+
 	render() {
 		return html`
 			<div role="list" class="d2l-list-container">
 				<slot></slot>
 			</div>
 		`;
 	}
+
+	selectionState() {
+		const checkedListItems = this.itemsChecked();
+		if (!checkedListItems || checkedListItems.length < 1) {
+			return selectableListStates.none;
+		}
+
+		const uncheckedListItems = this.querySelectorAll('d2l-list-item:not([checked])');
+		if (uncheckedListItems.length < 1) {
+			return selectableListStates.all;
+		}
+
+		return selectableListStates.indeterminate;
+	}
+
+	selectAll() {
+		if (!this.selectable) {
+			return;
+		}
+		const uncheckedListItems = this.querySelectorAll('d2l-list-item:not([checked])');
+		if (uncheckedListItems.length < 1) {
+			const checkedListItems = this.querySelectorAll('d2l-list-item[checked]');
+			checkedListItems.forEach(listItem => {
+				listItem.toggleAttribute('checked');
+			});
+		} else {
+			uncheckedListItems.forEach(listItem => {
+				listItem.toggleAttribute('checked');
+			});
+		}
+	}
+
+	_onlistItemChecked(event) {
+		event.stopPropagation();
+
+		this._checked[event.detail.ref] = event.detail.checked;
+		this.dispatchEvent(new CustomEvent('change', {
+			detail: {
+				checkedItems: this.itemsChecked()
+			}
+		}));
```

Comment:
Thinking about future components that might be interested in knowing these counts... I think ideally if you "select all" or "unselect all" it would only fire a single "change" event for the whole list. That way things depending on it would only re-render once. Thoughts?

Reply:
Yes, I would love for that to happen. Right now you get an event per item. I don't know exactly how to make this work but I will give it more thought.

Reply:
One idea would be to put a `setIsSelected(bool isSelected)` method on the items that you call instead of setting their attribute. It could take an optional second parameter `supressEvent` where by default it would fire an event but here you could pass in `true` to cause it not to.

Reply:
Love it!

- [ ] Add `setIsSelected` with optional `supressEvent`